### PR TITLE
osd/PrimaryLogPG: update missing_loc more carefully

### PR DESCRIPTION
**Pull request by xiexingguo**
There could be clones and hence we should update missing_loc
based on log entry itself.
    
Also I think the **have** field of missing_loc is not quite accurate here
but should be fine.

Fixes: http://tracker.ceph.com/issues/35546    
Signed-off-by: xie xingguo <xie.xingguo@zte.com.cn>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by liewegas** [comment]
We need a tracker bug for this since we'll need to backport the fix?

**Comment by xiexingguo** [comment]
> We need a tracker bug for this since we'll need to backport the fix?

Sure. http://tracker.ceph.com/issues/35546

**Comment by xiexingguo** [comment]
@jdurgin ping

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -10382,32 +10382,22 @@ void PrimaryLogPG::issue_repop(RepGather *repop, OpContext *ctx)
...
-    if (async_recovery_targets.count(peer) && peer_missing[peer].is_missing(soid)) {
-      for (auto &&entry: ctx->log) {
-	missing_loc.add_missing(soid, ctx->at_version, eversion_t(), entry.is_delete());
-      }
-    }
-  }
-
-  dout(30) << __func__ << " missing_loc before: " << missing_loc.get_locations(soid) << dendl;
-
   if (requires_missing_loc) {
```
Comment (neha-ojha) [inline src/osd/PrimaryLogPG.cc]: @xiexingguo This looks more like using `requires_missing_loc` to update `missing_loc`, which looks fine to me. Not sure, I understand the commit description "There could be clones and hence we should update missing_loc
based on log entry itself".

Comment (xiexingguo) [reply src/osd/PrimaryLogPG.cc]: No, that is not the key point.

File: src/osd/PrimaryLogPG.cc [hunk 2]
```diff
@@ -10382,32 +10377,22 @@ void PrimaryLogPG::issue_repop(RepGather *repop, OpContext *ctx)
...
   }
 
-  for (set<pg_shard_t>::const_iterator i = acting_recovery_backfill.begin();
-       i != acting_recovery_backfill.end();
-       ++i) {
-    pg_shard_t peer(*i);
-    if (peer == pg_whoami) continue;
-    if (async_recovery_targets.count(peer) && peer_missing[peer].is_missing(soid)) {
-      for (auto &&entry: ctx->log) {
-	missing_loc.add_missing(soid, ctx->at_version, eversion_t(), entry.is_delete());
```
Comment (xiexingguo) [inline src/osd/PrimaryLogPG.cc]: @neha-ojha 

If we modified a snapshot-object, we should clone it first. Hence there should be two log entries in the log list of opctx, one for the head object(whose object-id is soid), and the other is for the cloned object(whose object-id is **not** soid, it should be coid instead, see __PrimaryLogPG::make_writeable__). So we should update missing_loc by the log entry itself here, just like we update peer_missing.

**Comment by xiexingguo** [comment]
http://pulpito.ceph.com/xxg-2018-09-07_12:44:50-rados-wip-xxg-testing-2018-09-07-1720-distro-basic-smithi/

**Pull request merged**